vector/src/test/java/im/vector/app/features/poll/create/CreatePollViewModelTest.kt

```diff
@@ -0,0 +1,202 @@
+import im.vector.app.features.poll.create.CreatePollViewStates.fakeRoomId
+import im.vector.app.features.poll.create.CreatePollViewStates.initialCreatePollViewState
+import im.vector.app.features.poll.create.CreatePollViewStates.pollViewStateWithOnlyQuestion
+import im.vector.app.features.poll.create.CreatePollViewStates.pollViewStateWithQuestionAndEnoughOptions
+import im.vector.app.features.poll.create.CreatePollViewStates.pollViewStateWithQuestionAndMaxOptions
+import im.vector.app.features.poll.create.CreatePollViewStates.pollViewStateWithQuestionAndNotEnoughOptions
+import im.vector.app.features.poll.create.CreatePollViewStates.pollViewStateWithoutQuestionAndEnoughOptions
+import im.vector.app.test.test
+import io.mockk.every
+import io.mockk.mockk
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.test.runTest
+import org.amshove.kluent.shouldBe
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.matrix.android.sdk.api.session.Session
+import org.matrix.android.sdk.api.session.getRoom
+import org.matrix.android.sdk.api.session.room.Room
+import org.matrix.android.sdk.api.session.room.model.message.PollType
+import org.matrix.android.sdk.api.session.room.send.SendService
+import org.matrix.android.sdk.api.util.Cancelable
+
+class CreatePollViewModelTest {
+
+    @get:Rule
+    val mvrxTestRule = MvRxTestRule()
+
+    private val fakeSession = mockk<Session>()
+    private val fakeRoom = mockk<Room>()
+    private val fakeSendService = mockk<SendService>()
+    private val fakeCancellable = mockk<Cancelable>()
+
+    private fun createPollViewModel(pollMode: PollMode): CreatePollViewModel {
+        return if (pollMode == PollMode.EDIT) {
+            CreatePollViewModel(CreatePollViewState(editPollArgs), fakeSession)
+        } else {
+            CreatePollViewModel(CreatePollViewState(createPollArgs), fakeSession)
+        }
+    }
+
+    @Before
+    fun setup() {
+        every { fakeSession.getRoom(fakeRoomId) } returns fakeRoom
+        every { fakeRoom.sendService() } returns fakeSendService
+        every { fakeSendService.sendPoll(any(), fakeQuestion, any()) } returns fakeCancellable
+    }
+
+    @Test
+    fun `given the view model is initialized then poll cannot be created and more options can be added`() = runTest {
+        val createPollViewModel = createPollViewModel(PollMode.CREATE)
+        createPollViewModel
+                .test()
+                .assertState(initialCreatePollViewState)
+                .finish()
+    }
+
+    @Test
+    fun `given there is not any options when the question is added then poll cannot be created and more options can be added`() = runTest {
+        val createPollViewModel = createPollViewModel(PollMode.CREATE)
+        createPollViewModel.handle(CreatePollAction.OnQuestionChanged(fakeQuestion))
+
+        // We need to wait for createPollViewModel.onChange is triggered
+        delay(10)
```

Comment:
You can use `assertStatesChanges` instead of `assertState` to avoid having to add a delay which may lead to test result inconsistencies.

Reply:
different cpu/computers will be able to execute faster/slower, we should avoid relying on delays whenever possible :crossed_fingers: 

`// We need to wait for createPollViewModel.onChange is triggered` in the test enviroment we should control all the dispatching and block asynchronous calls (which is the purpose of `runTest` and `MvRxTestRule`)

Reply:
different cpu/computers will be able to execute faster/slower, we should avoid relying on delays whenever possible :crossed_fingers: 

`// We need to wait for createPollViewModel.onChange is triggered` in the test enviroment we should control all the dispatching and block asynchronous calls (which is the purpose of `runTest` and `MvRxTestRule`)

Reply:
I tried `assertStatesChanges` and verify with a timeout but no luck.

Reply:
the issue here looks like your `handle` is not blocking correctly :thinking: 

the `ViewModel.test()`extension is how we're capturing emissions from the `ViewModel`, if we interact with the `ViewModel` before `test()` then we'll miss state events

```kotlin
val test = createPollViewModel.test() // register state and event listeners

viewModel.handle()  // blocking - causes a state update

test.assert() // asserts against the captured states
```

however here I can see the `handle` emissions happening after `.test()`


```
println("start")
val createPollViewModel = createPollViewModel(PollMode.CREATE)
createPollViewModel.handle(CreatePollAction.OnQuestionChanged(fakeQuestion))
println("handle")

// We need to wait for createPollViewModel.onChange is triggered
delay(10)
createPollViewModel
        .also { println("test") }
        .test() // prints captures emissions
        .assertState(pollViewStateWithOnlyQuestion)
        .finish()
```

```
start
handle
test
CreatePollViewState(roomId=fakeRoomId, editedEventId=null, mode=CREATE, question=What is your favourite coffee?, options=[, ], canCreatePoll=false, canAddMoreOptions=true, pollType=DISCLOSED_UNSTABLE)
```
notice how the emission happens after `test()`, which shows the logic and updates inside `handle` are non blocking (a test's worst enemy!)

Reply:
separating the `test()` resolves the issue for me 

```
println("start")
val createPollViewModel = createPollViewModel(PollMode.CREATE)
val test = createPollViewModel.test()

createPollViewModel.handle(CreatePollAction.OnQuestionChanged(fakeQuestion))
println("handled")

test
        .also { println("test") }
        .assertStates(initialCreatePollViewState, pollViewStateWithOnlyQuestion)
        .finish()
```

```
start
CreatePollViewState(roomId=fakeRoomId, editedEventId=null, mode=CREATE, question=, options=[, ], canCreatePoll=false, canAddMoreOptions=true, pollType=DISCLOSED_UNSTABLE)
CreatePollViewState(roomId=fakeRoomId, editedEventId=null, mode=CREATE, question=What is your favourite coffee?, options=[, ], canCreatePoll=false, canAddMoreOptions=true, pollType=DISCLOSED_UNSTABLE)
handled
test
```

Reply:
I tried many approaches. The problem is not able to sync the `canCreatePoll` state field. To be clear: Could you please try to run `given there is a question when enough options are added then poll can be created and more options can be added` with your suggestion?

Reply:
I understood the problem: in the `test()` extension method, we use a different dispatcher to assert values on the flows than the one used in the `MvRxTestRule`. @onurays I can apply a fix if you want?

Reply:
I might be missing something here, there's no issue in the `test()` extension, the problems are 

- the usage of `.test()` should be before any interactions with the `ViewModel`
- `ViewModels` can emit multiple states, especially when multiple handles are called


```kotlin
@Test
fun `given there is not a question when enough options are added then poll cannot be created and more options can be added`() = runTest {
    val createPollViewModel = createPollViewModel(PollMode.CREATE)
    val test = createPollViewModel.test()

    repeat(CreatePollViewModel.MIN_OPTIONS_COUNT) {
        createPollViewModel.handle(CreatePollAction.OnOptionChanged(it, A_FAKE_OPTIONS[it]))
    }

    test.assertStatesChanges(
            initialCreatePollViewState,
            { copy(options = listOf(A_FAKE_OPTIONS[0], "")) },
            { copy(options = listOf(A_FAKE_OPTIONS[0], A_FAKE_OPTIONS[1])) },
    ).finish()
}

@Test
fun `given there is a question when enough options are added then poll can be created and more options can be added`() = runTest {
    val createPollViewModel = createPollViewModel(PollMode.CREATE)
    val test = createPollViewModel.test()

    createPollViewModel.handle(CreatePollAction.OnQuestionChanged(A_FAKE_QUESTION))
    repeat(CreatePollViewModel.MIN_OPTIONS_COUNT) {
        createPollViewModel.handle(CreatePollAction.OnOptionChanged(it, A_FAKE_OPTIONS[it]))
    }

    test.assertStatesChanges(
            initialCreatePollViewState,
            { copy(question = A_FAKE_QUESTION) },
            { copy(options = listOf(A_FAKE_OPTIONS[0], "")) },
            { copy(options = listOf(A_FAKE_OPTIONS[0], A_FAKE_OPTIONS[1])) },
    ).finish()
}
```

Reply:
the ordering is important because we're subscribing to the `ViewModel` flows in order to capture their emissions

if all the helper logic was inlined it would look like...

```kotlin
// .test()
val captures = mutableListOf()
viewModel.stateFlow
  .onEach { captures.add(it}
  .launchIn(nonBlockingContext)

// handle
viewModel.setState {
    copy("foo")
}

// assert
captures shouldBeEqualTo expectedEmissions
```

Reply:
turns out there's a bug/coroutine-test incompatibility  https://github.com/airbnb/mavericks/issues/599

```kotlin
@get:Rule
val mvrxTestRule = MvRxTestRule(testDispatcher = UnconfinedTestDispatcher())

@Test
fun `can create poll`() = runTest {
    val createPollViewModel = createPollViewModel(PollMode.CREATE)
    val test = createPollViewModel.test()

    createPollViewModel.handle(CreatePollAction.OnQuestionChanged(A_FAKE_QUESTION))
    repeat(CreatePollViewModel.MIN_OPTIONS_COUNT) {
        createPollViewModel.handle(CreatePollAction.OnOptionChanged(it, A_FAKE_OPTIONS[it]))
    }

    test.assertStatesChanges(
            initialCreatePollViewState,
            { copy(question = A_FAKE_QUESTION) },
            { copy(options = listOf(A_FAKE_OPTIONS[0], "")) },
            { copy(options = listOf(A_FAKE_OPTIONS[0], A_FAKE_OPTIONS[1])) },
            { copy(canCreatePoll = true) },
    ).finish()
}
```

Reply:
Thank you both for the great help! Please see my last commit (and approve :).